node/network/approval-distribution/src/lib.rs

```diff
@@ -89,6 +97,132 @@ impl RecentlyOutdated {
+			(false, true) => RequiredRouting::GridX, // messages from Y go to X
+			(true, true) => RequiredRouting::GridXY, // if the grid works as expected, this shouldn't happen.
+		}
+	}
+
+	// Get a filter function based on this topology and the required routing
+	// which returns `true` for peers that are within the required routing set
+	// and false otherwise.
+	fn route_to_peer(&self, required_routing: RequiredRouting, peer: &PeerId) -> bool {
+		match required_routing {
+			RequiredRouting::All => true,
+			RequiredRouting::GridX => self.peers_x.contains(peer),
+			RequiredRouting::GridY => self.peers_y.contains(peer),
+			RequiredRouting::GridXY => self.peers_x.contains(peer) || self.peers_y.contains(peer),
+			RequiredRouting::None | RequiredRouting::PendingTopology => false,
+		}
+	}
+}
+
+impl From<network_bridge_event::NewGossipTopology> for SessionTopology {
+	fn from(topology: network_bridge_event::NewGossipTopology) -> Self {
+		let peers_x =
+			topology.our_neighbors_x.values().flat_map(|p| &p.peer_ids).cloned().collect();
+		let peers_y =
+			topology.our_neighbors_y.values().flat_map(|p| &p.peer_ids).cloned().collect();
+
+		let validator_indices_x =
+			topology.our_neighbors_x.values().map(|p| p.validator_index.clone()).collect();
+		let validator_indices_y =
+			topology.our_neighbors_y.values().map(|p| p.validator_index.clone()).collect();
+
+		SessionTopology { peers_x, peers_y, validator_indices_x, validator_indices_y }
+	}
+}
+
+#[derive(Default)]
+struct SessionTopologies {
+	inner: HashMap<SessionIndex, (Option<SessionTopology>, usize)>,
+}
+
+impl SessionTopologies {
+	fn get_topology(&self, session: SessionIndex) -> Option<&SessionTopology> {
+		self.inner.get(&session).and_then(|val| val.0.as_ref())
+	}
+
+	fn inc_session_refs(&mut self, session: SessionIndex) {
+		self.inner.entry(session).or_insert((None, 0)).1 += 1;
+	}
+
+	fn dec_session_refs(&mut self, session: SessionIndex) {
+		if let hash_map::Entry::Occupied(mut occupied) = self.inner.entry(session) {
+			occupied.get_mut().1 = occupied.get().1.saturating_sub(1);
+			if occupied.get().1 == 0 {
+				let _ = occupied.remove();
+			}
+		}
+	}
+
+	// No-op if already present.
+	fn insert_topology(&mut self, session: SessionIndex, topology: SessionTopology) {
+		let entry = self.inner.entry(session).or_insert((None, 0));
+		if entry.0.is_none() {
+			entry.0 = Some(topology);
+		}
+	}
+}
+
+// A note on aggression thresholds: changes in propagation apply only to blocks which are the
+// _direct descendants_ of the finalized block which are older than the given threshold,
+// not to all blocks older than the threshold. Most likely, a few assignments struggle to
+// be propagated in a single block and this holds up all of its descendants blocks.
+// Accordingly, we only step on the gas for the block which is most obviously holding up finality.
+#[derive(Clone)]
+struct AggressionConfig {
+	/// Aggression level 1: all validators send all their own messages to all peers.
```

Comment:
Do we have a type for `BlockNumber`-`Delta`?

Reply:
I think measuring durations in block numbers is reasonable. It is a distance metric, after all.